Refactor: Move realtime stream processing outside the initial HTTP Request

leszko opened the pull request:
This is a refactor-only PR which moves the Realtime Stream processing logic outside of the HTTP Request.

So, now the flow is clearer:
1. We send the initial HTTP Request to Orchestrator (selection + send initial HTTP Req)
2. We start all the stream processing (Trickle channels, etc.)

leszko committed:
Refactor: Move realtime stream processing outside the initial HTTP Request

leszko committed:
Refactor: Inline onFirstSegment

j0sh commented on server/ai_live_video.go:
```diff
@@ -246,7 +247,22 @@ func startTrickleSubscribe(ctx context.Context, url *url.URL, params aiRequestPa
 			if firstSegment {
 				firstSegment = false
-				onFistSegment()
+				delayMs := time.Since(params.liveParams.startTime).Milliseconds()
```
excellent improvement

leszko replied:
Thanks! :)

j0sh commented on server/ai_mediaserver.go:
```diff
@@ -624,14 +624,77 @@ func (ls *LivepeerServer) StartLiveVideo() http.Handler {
 }
 
+func processStream(ctx context.Context, params aiRequestParams, req worker.GenLiveVideoToVideoJSONRequestBody) {
+	params.liveParams.outputWriter = startOutput(ctx, params)
```
I am a little confused if this is actually working right now? I can see it starts the ffmpeg goroutines but don't see where the `outputWriter` is wired in. Also wouldn't the old code setting up the output need to be removed?

BTW also have a soft preference for returning an error here and pushing that handling upstream instead of calling `stopPipeline` within the startOutput function, eg

```
params.liveParams.outputWriter, err = startOutput(ctx, params)
if err != nil {
  return err
}
```

leszko replied:
Ahh, yeah, right, I'll remove the old code + update as you suggested. Good points, thanks!

j0sh commented on server/ai_mediaserver.go:
```diff
@@ -624,14 +624,77 @@ func (ls *LivepeerServer) StartLiveVideo() http.Handler {
 	})
 }
 
+func processStream(ctx context.Context, params aiRequestParams, req worker.GenLiveVideoToVideoJSONRequestBody) {
```
Soft preference to push the error handling up as much as possible - if we return an error here then we can call stopPipeline just once at the call site

j0sh reviewed (approved):
Very good change, just one question about whether ffmpeg output is actually working, and some soft suggestions around error handling

leszko committed:
Refactor startOutput()

leszko committed:
Move writer to startOutput()

leszko committed:
Close outputWriter

leszko committed:
Move inputStreamExists() into ai_live_video.go

leszko committed:
Add startLivePipeline() to mark that the stream is started

Otherwise the stream output crashes because it thinks the input stream does not exists

leszko committed:
Revert writer change

leszko commented:
Thanks @j0sh for the review. Added one more change: https://github.com/livepeer/go-livepeer/pull/3565/commits/97000d575ec0b98de28c72c27277d80a153b3d73, so not merging today, to give you the change to re-review.

It's mainly because we need to mark that the input stream is live before we start the output. Otherwise in [this condition](https://github.com/livepeer/go-livepeer/blob/6fd12bbbfed12274d45028742915b369455eaab4/server/ai_live_video.go#L301) ffmpeg crashes before we start stream processing. PTAL.

leszko committed:
Merge branch 'master' into rafal/refactor-live-processing

j0sh commented:
> Added one more change: https://github.com/livepeer/go-livepeer/commit/97000d575ec0b98de28c72c27277d80a153b3d73, so not merging today, to give you the change to re-review.

Ah thanks for the heads up on this. Now I am starting to feel a little uncertain about the changes involved to support ffmpeg here; the control channel changes are getting a little ad-hoc, and we essentially recreate the LivePipelines entry once we have an orchestrator. It has always felt kind of awkward to use "control channel presence" as a proxy for whether the stream is actually connected, especially now since we may be connecting to ffmpeg without having an orchestrator (and hence no control channel) [1]

Since there are basically two changes in this PR anyway (ffmpeg + processAIRequest), maybe just drop the ffmpeg / control changes for now so the processAIRequest refactor can move forward. I have been working on [improving ffmpeg output](https://github.com/livepeer/go-livepeer/pull/3576) anyway and can take a stab at moving ffmpeg out once that is done.

 [1] I tried to kind of emphasize that the LivePipelines data is "just a control channel!" [here](https://github.com/livepeer/go-livepeer/pull/3502) and [here](https://github.com/livepeer/go-livepeer/pull/3503) so it is kind of funny to see the ~3rd rename for `cleanupLivePipelines` :) Maybe it's the LivePipelines that needs to be renamed...

leszko committed:
Revert "Add startLivePipeline() to mark that the stream is started"

This reverts commit 97000d575ec0b98de28c72c27277d80a153b3d73.

leszko commented:
> > Added one more change: [97000d5](https://github.com/livepeer/go-livepeer/commit/97000d575ec0b98de28c72c27277d80a153b3d73), so not merging today, to give you the change to re-review.
> 
> Ah thanks for the heads up on this. Now I am starting to feel a little uncertain about the changes involved to support ffmpeg here; the control channel changes are getting a little ad-hoc, and we essentially recreate the LivePipelines entry once we have an orchestrator. It has always felt kind of awkward to use "control channel presence" as a proxy for whether the stream is actually connected, especially now since we may be connecting to ffmpeg without having an orchestrator (and hence no control channel) [1]
> 
> Since there are basically two changes in this PR anyway (ffmpeg + processAIRequest), maybe just drop the ffmpeg / control changes for now so the processAIRequest refactor can move forward. I have been working on [improving ffmpeg output](https://github.com/livepeer/go-livepeer/pull/3576) anyway and can take a stab at moving ffmpeg out once that is done.
> 
> [1] I tried to kind of emphasize that the LivePipelines data is "just a control channel!" [here](https://github.com/livepeer/go-livepeer/pull/3502) and [here](https://github.com/livepeer/go-livepeer/pull/3503) so it is kind of funny to see the ~3rd rename for `cleanupLivePipelines` :) Maybe it's the LivePipelines that needs to be renamed...

Ok, I think it's a good idea. I moved ffmpeg back into `startTrickleSubscribe()`. Then, we can address it separately. I plan to merge this PR on Monday after deploy (unless you have any concerns).

leszko committed:
Move back ffmpeg out into startTrickleSubscribe()

leszko committed:
Merge remote-tracking branch 'origin/rafal/refactor-live-processing' into rafal/refactor-live-processing

leszko committed:
Merge branch 'master' into rafal/refactor-live-processing

j0sh reviewed (approved):
LGTM, thanks for doing this!

leszko committed:
Merge branch 'master' into rafal/refactor-live-processing

leszko committed:
Merge master in
